### ISSUE TITLE
Could I generate more than 1 character at the same time?

### ISSUE THREAD

**Issue by vinhqdang** [jcjohnson/torch-rnn#89]
Hello,

For instance I have a model already and a start text "abcd"

I would like to predict the next character, but instead of predict the 1 next character, I would like to see multiple potential characters which could be the next after "abcd".

For instance, something like:

a: 0.5
b:0.4
c:0.1
...

It is possible to do with torch-rnn?

Thanks a lot,

**Comment by jcjohnson** [comment]
In general RNN language models give a distribution over the next character conditioned on the previous characters, and torch-rnn is no different. The probability distribution over next characters is computed here:

https://github.com/jcjohnson/torch-rnn/blob/master/LanguageModel.lua#L195

There's no existing script or function to pull out or save these distributions, but it would be pretty easy to implement.

**Comment by vinhqdang** [comment]
Hello

Thanks a lot, your answer is great. However, i have some more questions:

1. When I print ``probs``, I got a table of ``n+1`` values, with ``n`` is the number of different character appeared in my training set. How should I interpret the table? For instance, the first value is 1.47, what is the corresponding character? Also, the last value is 0.12, how could I know the corresponding character?

2. When I tried, I observed that ``probs`` for the same ``start_text`` are always the same. Why does the generated character may differ over time?

Thanks

**Comment by robinsloan** [comment]
Hey Vinh:

1. The probability's order is the same as your vocabulary. You could determine which character that first probability value of 1.47 corresponds to using `self.idx_to_token[1]`. Likewise, if the last value of 0.12 is at index (say) 36, use `self.idx_to_token[36]`. (Also: the `probs` table is exactly as large as your vocabulary. I suspect the `+1` you're seeing is the space character. Maybe?)

2. The `probs` are always the same, but samples can be different because the `multinomial` function (which you are probably using) truly *samples*, rather than just choosing the index with the highest probability. Think of it this way: the roulette wheel never changes, but that doesn't mean you get the same result every time you spin it :)

And, if I'm interpreting your original question correctly, you can use something like this:

```
local probs = torch.div(scores, temp):double():exp():squeeze()
probs:div(torch.sum(probs))
local num_samples = 5
local a_bunch_of_possible_chars = torch.multinomial(probs, num_samples, false)
```

That last `false` tells `torch.multinomial` to choose a different index each time, which I find is usually the behavior I want.

**Comment by vinhqdang** [comment]
Hi @robinsloan 👍

**Comment by vinhqdang** [comment]
Hello @robinsloan 

Sorry that I have to open this issue again.

I noticed that, the distribution of next character in my case does not change with different ``start_text``.

For instance, if I have a model X.t7, then I want to see the distribution of next potential character with two different start_text "abcd" and "ghik", but it seems to me that the distribution, i.e. ``self.idx_to_token[1]`` to ``self.idx_to_token[5]`` always be the same.

Is it normal, or I did some critical mistakes?

Thanks

**Comment by robinsloan** [comment]
Hey there -- I presume you're "warming up" the network with the start text as in the [`sample` code here](https://github.com/jcjohnson/torch-rnn/blob/master/LanguageModel.lua#L170)? The crucial step is on line #177. Also: be sure you are saving the return value from your `self:forward` call into a `scores` variable -- that's what you use to calculate the new probabilities each time!

Here's a simple test -- drop this function somewhere inside LanguageModel.lua and see if you can validate that it's working:

```
function LM:test()
  self:resetStates()

  local x = self:encode_string("FOO"):view(1, -1)
  local T0 = x:size(2)
  scores = self:forward(x)[{{}, {T0, T0}}]
  local probs = torch.div(scores, 1.0):double():exp():squeeze()
  probs:div(torch.sum(probs))
  local next_char = torch.multinomial(probs, 1):view(1, 1)
  print("got next char index for FOO:")
  print(next_char)
  -- print(probs)

  self:resetStates()

  x = self:encode_string("JAG"):view(1, -1)
  T0 = x:size(2)
  scores = self:forward(x)[{{}, {T0, T0}}]
  probs = torch.div(scores, 1.0):double():exp():squeeze()
  probs:div(torch.sum(probs))
  next_char = torch.multinomial(probs, 1):view(1, 1)
  print("got next char index for JAG; it should be different!")
  print(next_char)
  -- print(probs)
end
```

If not: there is indeed something wrong, and I'm afraid it's on a deeper level than I can diagnose.

**Comment by vinhqdang** [comment]
So, the key is ``self:resetStates()``?

But, as I saw in line #202, the function is called already before the function ends? What is the difference?

**Comment by robinsloan** [comment]
I wouldn't say `self:resetStates()` is "the key"… in fact I don't think it's your problem -- forgetting to call it would result in mysteriously *variable* output, not mysteriously *consistent* output.

**Comment by vinhqdang** [comment]
I saw the problem maybe:

If I use ``torch.multinomial``, sure, the ``next_char`` will be different, even for same ``start_text``.

However, my problem is:

For each ``start_text``, I want to extract next ``k`` most potential character, therefore I do:

```
for i=1,k do
         n_char = self.idx_to_token[i]
         ...
end
```

And I observe that, for any ``start_text``, self.idx_to_token[1], self.idx_to_token[2] .... will be the same.

So, the point now is, is it expected?

**Comment by robinsloan** [comment]
Hmm -- the important thing, how are you generating those indexes? In your (pseudo)code above, you'll get the same results every time because you're asking the model for the characters corresponding to 1, 2, 3, 4, 5... every time. So the crucial part comes before the (pseudo)code you've listed.

Try something like this:

```
function LM:test()
  self:resetStates()

  local x = self:encode_string("FOO"):view(1, -1)
  local T0 = x:size(2)
  scores = self:forward(x)[{{}, {T0, T0}}]
  local probs = torch.div(scores, 1.0):double():exp():squeeze()
  probs:div(torch.sum(probs))
  local next_chars, next_indices = torch.topk(probs, 10)
  print("got next 10 char indices for FOO:")
  print(next_indices)
  -- print(probs)

  self:resetStates()

  x = self:encode_string("JAG"):view(1, -1)
  T0 = x:size(2)
  scores = self:forward(x)[{{}, {T0, T0}}]
  probs = torch.div(scores, 1.0):double():exp():squeeze()
  probs:div(torch.sum(probs))
  next_chars, next_indices = torch.topk(probs, 10)
  print("got next 10 char indices for JAG; it should be different!")
  print(next_indices)
  -- print(probs)
end
```

Again: if you run that (or something like it) and print out exactly the same `next_indices` both times, you'll know something is definitely wrong.

([Docs on torch.topk.](https://github.com/torch/torch7/blob/master/doc/maths.md#torchtopkresval-resind-x-k-dim-dir-sort))

**Comment by vinhqdang** [comment]
Oh s***, it is exactly what happen. Two next_indices are the same

**Comment by robinsloan** [comment]
Hahahaha oh no! Maybe try a different model? I have a pre-trained one [here](http://from.robinsloan.com/rnn-writer/scifi-model.zip).

But honestly, I don't know. You have officially moved beyond my small sphere of competence. GOOD LUCK SIR! 😉

**Comment by vinhqdang** [comment]
Hi @robinsloan 

So sorry, I have to come back this problem one more time :(

So, I trained the model again. This time everything is okay: different start_text, different probs value.

However, I have a question with ``topk`` function (sorry, I am a newbie on Lua).

In the code, I did:

```{lua}
print probs
local next_chars, next_indices = torch.topk(probs, 5)
print (next_indices)
```

and in terminal I have:

```
 0.1377
 0.3687
 0.4707
 0.6819
 0.7766
 0.4042
 0.8941
 0.2265
 0.1577
 0.5926
 0.1547
[torch.DoubleTensor of size 11]

  1
 11
  9
  8
  2
[torch.LongTensor of size 5]
```

(there are 10 different characters in my train set + newline character)

It is not what I really expect. From ``probs`` we could see that the maximum value is 0.8491 (line 7), then 0.7766 (line 5) etc.

So from probs, I want to export 5 indexes of highest probability, i.e. it should be in this case
7 5 4 10 3

**Comment by robinsloan** [comment]
@vinhqdang, I totally led you astray! By default, `topk` (for reasons that I'm sure are good, but inscrutable to me) returns the *smallest* `n` elements in the array. The arguments are `torch.topk(x, k, dim, dir)` so you can reverse this behavior and get the results you want using this:

```Lua
y, i = torch.topk(probs, 5, 0, true)
```

There are some other nuances to the function that might trip you up, so I definitely recommend [reading the documentation](https://github.com/torch/torch7/blob/master/doc/maths.md#torchtopkresval-resind-x-k-dim-dir-sort).

**Comment by vinhqdang** [comment]
Thanks a lot,

So the correct one is (I think)

```{lua}
local next_chars, next_indices = torch.topk(probs,  5, 1,true, true)
```

**Comment by robinsloan** [comment]
Note that you can elide the `dims` argument and just use `torch.topk(which_tensor, how_many, true)` though... I am not really sure why... so maybe using the verbose version is a better choice 😎

**Comment by vinhqdang** [comment]
I don’t know too, I will check. I should really learn Lua from now on.

**Comment by vinhqdang** [comment]
I don't think I can eliminate the last parameter

If I use

```{lua}
torch.topk(probs,  5, true)
```

It will print ``3 10 4 5 7`` (still 5 largest indexes, but not sorted)

**Comment by robinsloan** [comment]
Ah there you go! For my purposes I didn't need the `topk` sorted, but sounds like you should use the long version. I hope it works for you!

And, as for Lua: yeah, the fact that it's so "simple" makes it attractive to learn exactly as much as you need to solve a problem, and no more -- a behavior which has definitely come back around to haunt me.
